soc/st/stm32/common/stm32_iocell.c

```diff
@@ -0,0 +1,248 @@
+		return (FLASH->OBW1SR & FLASH_OBW1SR_XSPI1_HSLV) == FLASH_OBW1SR_XSPI1_HSLV;
+	} else if (domain == STM32_DT_IOCELL_XSPI2) {
+		return (FLASH->OBW1SR & FLASH_OBW1SR_XSPI2_HSLV) == FLASH_OBW1SR_XSPI2_HSLV;
+	} else {
+		return -EINVAL;
+	}
+#elif defined(CONFIG_SOC_SERIES_STM32N6X)
+	static uint32_t hconf1 = 0xFFFFFFFF;
+
+	if (hconf1 == 0xFFFFFFFF) {
+		WRITE_REG(BSEC->OTPCR, 124);
+
+		while ((BSEC->OTPSR & BSEC_OTPSR_BUSY) != 0U) {
+		}
+
+		if (BSEC->OTPSR & (BSEC_OTPSR_DISTURBF | BSEC_OTPSR_DEDF | BSEC_OTPSR_AMEF)) {
+			/* Read error => force safe configuration */
+			hconf1 = 0x0FFFFFFF;
+		} else {
+			hconf1 = BSEC->FVRw[124];
+		}
+	}
+
+	if (domain == STM32_DT_IOCELL_VDD) {
+		return (hconf1 & BIT(17)) != 0;
+	} else if (domain == STM32_DT_IOCELL_VDDIO2) {
+		return (hconf1 & BIT(16)) != 0;
+	} else if (domain == STM32_DT_IOCELL_VDDIO3) {
+		return (hconf1 & BIT(15)) != 0;
+	} else if (domain == STM32_DT_IOCELL_VDDIO4) {
+		return (hconf1 & BIT(14)) != 0;
+	} else if (domain == STM32_DT_IOCELL_VDDIO5) {
+		return (hconf1 & BIT(13)) != 0;
+	} else {
+		return -EINVAL;
+	}
+#endif /* CONFIG_SOC_SERIES_... */
+}
+
+/* Returns 0 on success
+ * Returns < 0 on error
+ */
+static inline int iocell_enable_hslv_runtime(uint16_t domain)
+{
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	if (domain == STM32_DT_IOCELL_VDDIO) {
+		LL_SBS_EnableIOSpeedOptim();
+	} else if (domain == STM32_DT_IOCELL_XSPI1) {
+		LL_SBS_EnableXSPI1SpeedOptim();
+	} else if (domain == STM32_DT_IOCELL_XSPI2) {
+		LL_SBS_EnableXSPI2SpeedOptim();
+	} else {
+		return -EINVAL;
+	}
+	return 0;
+#elif defined(CONFIG_SOC_SERIES_STM32N6X)
+	if (domain == STM32_DT_IOCELL_VDDIO2) {
+		LL_PWR_SetVddIO2VoltageRange(LL_PWR_VDDIO_VOLTAGE_RANGE_1V8);
+	} else if (domain == STM32_DT_IOCELL_VDDIO3) {
+		LL_PWR_SetVddIO3VoltageRange(LL_PWR_VDDIO_VOLTAGE_RANGE_1V8);
+	} else if (domain == STM32_DT_IOCELL_VDDIO4) {
+		LL_PWR_SetVddIO4VoltageRange(LL_PWR_VDDIO_VOLTAGE_RANGE_1V8);
+	} else if (domain == STM32_DT_IOCELL_VDDIO5) {
+		LL_PWR_SetVddIO5VoltageRange(LL_PWR_VDDIO_VOLTAGE_RANGE_1V8);
+	} else if (domain == STM32_DT_IOCELL_VDD) {
+		LL_PWR_SetVddIOVoltageRange(LL_PWR_VDDIO_VOLTAGE_RANGE_1V8);
+	} else {
+		return -EINVAL;
+	}
+	return 0;
+#endif /* CONFIG_SOC_SERIES_... */
+}
+
+static inline void iocell_enable_compensation(uint16_t domain)
+{
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	/* TODO: check MCU revision */
+
+	/* This implements workaround described in ES0596:
+	 * 2.2.15. I/O compensation could alter duty-cycle of high-frequency output signal
+	 */
+	uint32_t shift;
+	uint32_t nmos, pmos;
+
+	if (domain == STM32_DT_IOCELL_VDDIO) {
+		shift = 0;
+	} else if (domain == STM32_DT_IOCELL_XSPI1) {
+		shift = 1;
+	} else if (domain == STM32_DT_IOCELL_XSPI2) {
+		shift = 2;
+	} else {
+		return;
+	}
+	/* TODO: skip if compensation cell already enabled ? */
+
+	/* Configure Cell selection for domain and enable */
+	MODIFY_REG(SBS->CCCSR, BIT((shift * 2) + 1), BIT(shift * 2));
```

Comment:
1. Use the helpers
```suggestion
	stm32_reg_modify_bits(SBS->CCCSR, BIT((shift * 2) + 1), BIT(shift * 2));
```
2. Maybe the comment could be expanded a bit to indicate that the `mask` and `value` are unrelated (usually, `MODIFY_REG` is used to replace a field's value, whereas here it's used to clear one bit and set another at the same time, which is slightly surprising/confusing)